qiskit/pulse/schedule.py

```diff
@@ -1504,6 +1925,33 @@ def _get_timeslots(schedule: "ScheduleComponent") -> TimeSlots:
     return timeslots
 
 
+def _get_references(block_elms: List["BlockComponent"]) -> Set[Reference]:
+    """Recursively get reference instructions in the current scope.
+
+    Args:
+        block_elms: List of schedule block elements to investigate.
+
+    Returns:
+        A set of unique reference instructions.
+    """
+    references = []
+    for elm in block_elms:
+        if isinstance(elm, ScheduleBlock):
+            references += _get_references(elm._blocks)
+        elif isinstance(elm, Reference):
+            references.append(elm)
+    return set(references)
+
+
+def _scope_parameter(param: Parameter, scope: str, delimiter: str = ":") -> Parameter:
+    """Override parameter object with program scope information."""
+    new_name = f"{scope}{delimiter}{param.name}"
+    scoped_param = Parameter.__new__(Parameter, new_name, uuid=getattr(param, "_uuid"))
+    scoped_param.__init__(new_name)
+
+    return scoped_param
```

Comment:
This looks a bit terrifying to me, to be honest.  I get what you're doing, but it feels potentially very fragile.  I don't have a sensible alternative suggestion, though, so it's probably best to leave it.

Reply:
This is identical trick used in qpy loader, but I agree this is bit hackey. I really wanted to introduce dedicated parameter subclass for pulse module.